keps/sig-apps/0753-sidecarcontainers.md

```diff
@@ -109,34 +180,173 @@ section][#graduation-criteria].
+
+#### Logging/Metrics sidecar
+
+A logging sidecar should start before several other containers, to not lose logs
+from the startup of other applications. Let's call _main container_ the app that
+will log and _logging container_ the sidecar that will facilitate it.
+
+If the logging container starts after the main app, some logs can be lost.
+Furthermore, if the logging container is not yet started and the main app
+crashes on startup, those logs can be lost (depends if logs go to a shared volume
+or over the network on localhost, etc.). While you can modify your application
+to handle this scenario during startup (as it is probably the change you need to
+do to handle sidecar crashes), for shutdown this approach won't work.
+
+On shutdown the ordering behaviour is arguably more important: if the logging
+container is stopped first, logs for other containers are lost. No matter if
+those containers queue them and retry to send them to the logging container, or
+if they are persisted to a shared volume. The logging container is already
+killed and will not be restarted, as the pod is shutting down. In these cases,
+logs are lost.
+
+The same things regarding startup and shutdown apply for a metrics container.
+
+Some work-arounds can be done, to alleviate the symptoms:
+ * Ignore SIGTERM on the sidecar container, so it is alive until the pod is
+   killed. This is not ideal and _greatly_ increases the time a pod will take to
+terminate. For example, if the P99 response time is 2 minutes and therefore the
+TerminationGracePeriodSeconds is set to that, the main container can finish in 2
+seconds (that might be the average) but ignoring SIGTERM in the sidecar
+container will force the pod to live for 2 minutes anyways.
+ * Use preStop hooks that just runs a "sleep X" seconds. This is very similar to
+   the previous item and has the same problems.
+
+#### Service mesh
+
+Service mesh presents a similar problem: you want the service mesh container to
+be running and ready before other containers start, so that any inbound/outbound
+connections that a container can initiate goes through the service mesh.
+
+A similar problem happens for shutdown: if the service mesh container is
+terminated prior to the other containers, outgoing traffic from other apps will be
+blackholed or not use the service mesh.
+
+However, as none of these are possible to guarantee, most service meshes (like
+Linkerd and Istio), need to do several hacks to have the basic
+functionality. These are explained in detail in the
+[alternatives](#alternatives) section. Nonetheless, here is a  quick highlight
+of some of the things some service mesh currently need to do:
+
+ * Recommend users to delay starting their apps by using a script to wait for
+   the service mesh to be ready. The goal of a service mesh to augment the app
+   functionality without modifying it, that goal is lost in this case.
+ * To guarantee that traffic goes via the services mesh, an initContainer is
```

Comment:
Good point, yes, none of these are addressed (sorry, I wasn't aware that was an important problem). It is [a non-goal][non-goal] currently to run initContainer with sidecars containers.

The third option you mention, besides capabilities needed for the redirect, will not run concurrent with initContainers, so traffic is either blackholed or not, but initContainers traffic is not going through the istio sidecar either. 

The question is, then, do we want to have initContainers with sidecar containers in-scope or not? What do you think @howardjohn ?

The trade-off, from my POV, is:
 * Having it in-scope for this will solve/improve that problem (making this KEP way more complicated to move forward) 
vs
 * Having it out of scope will not solve that problem but will make advances with the rest

IMHO, it makes sense to move forward with this proposal (i.e. not solve the sidecars running concurrently with `initContainers`) but worries me a little that, if we later want to solve the initContainers with sidecar problem, the API might end up messy.

If we implement this KEP and in the future want to extend this in some way to address it, we can _technically_ do it using other container `types`. For example, `type: daemon` (probably a better name :-D) can be used to mean that (those containers starts before initContainers too). Or `type: OrderedPhase` with an `Order: <int>` and, let's say, using `Order: -1` can mean "before initContainers". However, it starts to get messy if you have containers in the `Containers` array that start before initContainers. 

So, that seems like a messy thing to me. If we want to go the initContainer+sidecars route, we might want to have a new `sidecarContainers` array that start before initContainers too. That was discussed in the past (only the array, I think, not starting before initContainers IIRC), it is mentioned in the [alternatives section][alternative] and why it was discarded, though.

At the same time, I'm not sure there are many sidecar types that might want to run concurrent with initContainers. Are there any other use case? Maybe if you want to collect metrics/logs from the the initContainers using a sidecar? Not sure there are many users for this, though. Although, maybe it doesn't matter/hurt to most of sidecars. What do others think?

Taking a step back, a question from someone out of the service mesh world: does the service mesh really wants to be part of the pod, if the life-cycle it needs is so different? I guess probably yes for latency and to easily transparently redirect traffic that might not be possible otherwise. But just making sure there are no other solutions and we really want to think about this problem when the service mesh lives in the pod :) 

Trying to be pragmatic, we can move to alpha stage without "initContainers + sidecars" (i.e. as it is proposed now) and gather feedback from users. We can gather feedback to see if it is worth doing a `sidecarContainers` array (or any other alternative that seems best to address the feedback)?

What do others think?

[non-goal]: https://github.com/kubernetes/enhancements/pull/1913/files#diff-c6212b56619f2b462935ad5f631d772fR348
[alternative]: https://github.com/kubernetes/enhancements/pull/1913/files#diff-c6212b56619f2b462935ad5f631d772fR1356

Reply:
If we implemented the KEP as is I would be very happy, it would improve a lot of issues for a lot of people. I would be even *more* happy if we also fixed the init container issue, but not at the expense of not fixing the startup/shutdown. no need to let perfect be the enemy of good.

It would probably make sense to future proof this so it *can* be done, but not worry about it right now. Having a new `type` seems feasible to me.

As far as other use cases - all service meshes require this, as far as I know, and logging seems like something you would want for init containers as well. Honestly most sidecars are feasible to run with the init containers, for all the same reasons you want them to run with the main container. For example, some people have SQL proxies - same is needed if you use SQL in init container.

> Taking a step back, a question from someone out of the service mesh world: does the service mesh really wants to be part of the pod, if the life-cycle it needs is so different

I don't think we *need* it to run in the pod, but there are no other existing alternatives that would work.

Reply:
Commenting from other thread on 

>  It was nice to have, but not many real use cases nor a big deal.

This claim is basically saying init containers that do networking is not a real use case, which isn't true in my opinion. There are very clear use cases for doing so, otherwise init containers would not likely exist :slightly_smiling_face:. I think its more accurate to say that the people you talked to happen to not have users complain about them much, but I can't imagine that generalizes well to the broader Kubernetes community

Reply:
Thanks! I agree that perfect is the enemy of good, I'm open to suggestions if sidecars make more sense before initContainers. But yeah, IMHO, the best solution might be not so "theoretical": implement either one of them and gather user feedback :-)

@derekwaynecarr What do you think?

Reply:
> I think its more accurate to say that the people you talked to happen to not have users complain about them much

This is exactly what I wanted to say. Sorry if it wasn't clear :)

Reply:
what will change in this KEP if the only change we will introduce is to start sidecar containers before init containers? Logging also fits nicely in this pattern.

Reply:
@SergeyKanzhelev the biggest change is semantics and how we want to reflect that, IMHO. Furthermore, those are not sidecars as we know them today, so can be somewhat unexpected. For example, some things we may need to decide, from the top of my head:

1. Will we use the current `containers` array for that? In that case, the `initContainers` array will start before the containers not type `sidecar` only (they are meant to start before any other container in the pod, the name `initContainers` might become confugsing. We might be able to clarify with docs). That might be unexpected, so we might want to consider using a `sidecarContainers` array to avoid confusions. Not really sure, something to sleep on it.
1. Today sidecar containers (as a concept, not the type as it is not repesnt) start after `initContainers` have run. If someone is using the `initContainers` to initialize something for sidecar containers (totally legit), another way to provide that functionality might be needed. Another `initSidecarContainers` thingy or something else? This can become weird
1. What to do if a sidecar crashes while some initContainers are running? Just restarts, I guess?

Thinking about this, I have mixed feelings, but I think I lean towards improving sidecars as we know them: sidecars that start after `initContainers` and use the alpha phase to gather feedback and revisit that. If we see that, for example, both things make sense, we might want to add a different type that starts before initContainers.

What do you think?

But I'm very open to thoughts :)

Reply:
*sorry for delayed response. Somehow all notifications broke for me. It's great to have a ping at a meeting.*

I though about these concerns. I understand the main concern is that sidecar containers that has a dependency on init containers would not work as expected when one simply adds a new type to the container? The assumption that everything will just work once conatiner marked as sidecar, may not be true. What if existing sidecar container has some synchronization logic that checks for payload to start and keeps it in "not started" state. This will break the entire pod initialization. 

The same time both important scenarios - logs and mesh would need an init containers support. If we will conclude that the way to support init containers is to introduce a similar type field on a container, why wouldn't we do it now. Sidecars which needs to work before init containers will just work. Sidecars which needs to work after init container may wait for the first readiness probe and treat it as an indication that initialization was complete.

Same as @howardjohn I believe that this KEP is great in addressing many issues. I just want to make sure there is no quick win by simply changing startup order for sidecars to run before init. And figuring out details like what to do when sidecars crashed during the initialization.

Perhaps there may be a section in the doc saying why we do not want to run sidecars before init.

Reply:
> _sorry for delayed response. Somehow all notifications broke for me. It's great to have a ping at a meeting._

Ups, thanks! :)

> I though about these concerns. I understand the main concern is that sidecar containers that has a dependency on init containers would not work as expected when one simply adds a new type to the container? The assumption that 

Yes, exactly, that is one concern I raised.

> everything will just work once conatiner marked as sidecar, may not be true. What if existing sidecar container has some synchronization logic that checks for payload to start and keeps it in "not started" state. This will break the entire pod initialization.

Not sure if we are saying the same here. That is exactly my point, if the sidecar has a dependency on an initContainer doing something before it starts, it will be completely broken.

> The same time both important scenarios - logs and mesh would need an init containers support. If we will conclude that the way to support init containers is to introduce a similar type field on a container, why wouldn't we do it now.

I think both options can work: we can do it now or we can have an alpha without initContainers, see if what we did effectively solves the problems that we think it solves or needs something else. After that, try to expand to `initContainers` and just focus on the problems of doing that, as we know what we have mostly works and the problems of extending it to that are quite different.

I lean towards extending into sidecars + initContianer after first alpha if we can't find something obvious that will work in the next few days. 

What do you think? 

> Sidecars which needs to work before init containers will just work.

No, they won't on all cases. This is why it is not trivial and I'd rather focus on one problem at a time, and gather feedback.

Istio is one example, they mention in this thread that the service mesh sidecar should be running before initContainers but that itself needs some initialization first. [They said][link]:

> A third option is potentially running the iptables in the actual sidecar, but this is not feasible due to permissions.

To add some more context: the problem they refer to is the permissions needed in the container to run the iptables rules. They don't want to have those permissions on the sidecar pod, and instead have them in the initContainer or CNI plugin.

So, they either need the CNI plugin to work in this case (for the iptables thing) or an initContainer that runs before sidecars, to run the iptables rules, then start the sidecar and then start the rest of the initContainer and continue with the initialization.

And Istio has a way out because they have the CNI plugin option, as they only want some iptables rules. That is, of course, not the case of all sidecars AFAIK. Any sidecar relying on an initContainer won't work if started before initContainers, even if they want to run during initContainer phase.

Some hacks can be done, like auto-inject the initContainer to be run first). But that has also several problems. Because then several applications want to be "the first" and there is a race injecting there. The analogous is alread happening to inject the last initContainers. Here is an example of such a bug report (it is also in some part of the text this PR adds :)): https://github.com/linkerd/linkerd2/issues/4758#issuecomment-658896063.

> Sidecars which needs to work after init container may wait for the first readiness probe and treat it as an indication that initialization was complete.

Not sure what you mean exactly with this, can you please elaborate? Wait for which readiness? How that would be explicit in the pod spec?

>  I just want to make sure there is no quick win by simply changing startup order for sidecars to run before init

IMHO, it doesn't seem like a quick win, but a path to explore that will take time. I think people will be opinionated, with very valid reasons, about having containers in the `containers` array starting before `initContainers` vs creating a new array (`sidecarContainers`), etc. 

The more we talk about it, the more I think it will be just easier if we keep the current scope for the alpha phase (sidecars don't run during initContainers), see if the mechanism we add for that are effective for those use cases and then focus on extending this to initContainers (we can make it a show-stopper for beta graduation if that makes anyone more comfortable).

What do you think?

[link]: https://github.com/kubernetes/enhancements/pull/1913#discussion_r462571359

Reply:
> Istio is one example, they mention in this thread that the service mesh sidecar should be running before initContainers but that itself needs some initialization first.

It would actually work fine I think. The sidecar does *not* need iptables set up. Our only requirement is the iptables *must* be set up before the app container. So:

* init, then app/sidecar in parallel is ok (but not ideal - hence the kep)
* sidecar, then init, then app is ok
* sidecar, then app, then init is not ok

I can see *hypothetical* scenarios that do require init containers to run first though, like if we use init containers to provision certificates. It could be worked around though, for example have the sidecar get ready without the certs and have the init container poll to ensure it doesn't exit until the sidecar has picked up the certs. Note this is still a much better scenario because it involves the sidecar/init container to make a change (2 images, likely owned by Istio/Linkerd/<ServiceMesh>) rather than every single application container.

Reply:
@rata I think we are on the same page understanding pros and cons of init and sidecars running in parallel. I think we prioritize different things while weighting options. So my questions is mostly to understand your priorities and understand the complexity. 

The way I see it - sidecar containers are something (as you mentioned earlier) that almost like "outside the pod" feature that represent an ambient state of a pod. Something that always exists and does not block termination. So developer should not think of synchronizing with it. And sidecar owner has guarantees that all app logic will be executed when sidecar is active. And I see init and regular containers as a developer's created app logic.

If developer wants to create it's own sidecar that is an app logic (like queue management or config synchronization), and it depends on some Init containers, this will be a little bit more involved as it will require a bit of synchronization with Init containers. I don't see the need for this synchronization implementation as a big deal. Today sidecars like this has are probably implementing even more logic.

As for semantic - we can call this KEP "daemon" or "ambient" containers instead of "sidecar" if this will make semantic clearer. 

Finally for the implementation complexity - are you talking about kubelet logic complexity or complexity for end user?

I also pinged you on Slack in case you want to discuss it in-person before writing more text =).